python/ray/tune/ray_trial_executor.py

```diff
@@ -256,23 +291,47 @@ def _setup_remote_runner(self, trial):
             logger.debug("Cannot reuse cached runner {} for new trial".format(
                 self._cached_actor))
             with self._change_working_directory(trial):
-                pg = self._pg_manager.clean_trial_placement_group(trial)
-
                 self._trial_cleanup.add(
-                    trial, actor=self._cached_actor, placement_group=pg)
+                    trial, actor=self._cached_actor, placement_group=None)
             self._cached_actor = None
 
-        _actor_cls = _class_cache.get(trial.get_trainable_cls())
+        trainable_cls = trial.get_trainable_cls()
+        if not trainable_cls:
+            raise AbortTrialExecution(
+                f"Invalid trainable: {trial.trainable_name}. If you passed "
+                f"a string, make sure the trainable was registered before.")
+        _actor_cls = _class_cache.get(trainable_cls)
+
         if trial.uses_placement_groups:
-            if not self._pg_manager.has_ready(trial.placement_group_factory):
+            if not self._pg_manager.has_ready(trial, update=True):
                 if trial not in self._staged_trials:
-                    if self._pg_manager.stage_trial_pg(
-                            trial.placement_group_factory):
+                    if self._pg_manager.stage_trial_pg(trial):
                         self._staged_trials.add(trial)
+                        self._just_staged_trials.add(trial)
+
+                just_staged = trial in self._just_staged_trials
+
+                if self._wait_for_pg is not None and (
+                        just_staged or self._trial_just_finished_before
+                        or not self.get_running_trials()):
+                    logger.debug(
+                        f"Waiting up to {self._wait_for_pg} seconds for "
+                        f"placement group of trial {trial} to become ready.")
+                    wait_end = time.monotonic() + self._wait_for_pg
+                    while time.monotonic() < wait_end:
+                        self._pg_manager.update_status()
+                        if self._pg_manager.has_ready(trial):
+                            break
+                        time.sleep(0.1)
```

Comment:
hmm, this is weird, is there any way to avoid this wait?

Reply:
(i actually don't understand what's going on in this block of code, would be great to get some clarity here!)

Reply:
I added a comment paragraph to the code. This is mostly for testing purposes. Generally for tests we would like to start a trial as soon as its resources are ready without calling `step()` another time (as it will be a non-deterministic number of times we would have to call `step()`, which is bad for the current test setups). So we have to wait until the placement group is ready, so we can start the remote actor on it. 

The number of seconds we wait is given by `self._wait_for_pg`. However, if we only used that we would always wait for, say, 5 seconds each time we try to start a trial. This is bad though, because in some cases we actually do have limited resources and the trial cannot be started until other trials finished. This will then dramatically slow down result handling (and in our cases, test execution time). So we only want to wait when it is likely that we will have resources available soon. This is the case when a) we just staged the trial (which is usually at the start of the run, or after another trial finished), b) when another trial actually finished (if the trial is already staged, but the resources only become available now), or c) when there are no running trials (maybe a trial errored and we're just waiting for the resources for the final trial).

This is also the reason for the `trial_just_finished` variables below

Reply:
> So we have to wait until the placement group is ready, so we can start the remote actor on it.

Is it possible to start the remote actor with the placement group before the PG is ready? Would that simplify things?